File: Sorting_Algorithm/Selection_Sort/selectionsort.java
package Sorting_Algorithm.Selection_Sort;

/**
 * selectionsort
 */
public class selectionsort {

    public static void main(String[] args) {
       int arr[]={9,6,2,1,4};
       System.out.println("Array before sorted=");
       for(int i:arr)
       {
        System.out.println(i);
       }
       selectionsort1(arr);
       System.out.println("Sorted elements are=");
       for(int i:arr)
       {
        System.out.println(i);
       }
       

    }
    public static void selectionsort1(int arr[])
    {
        for(int i=0;i<arr.length-1;i++)
        {
            int index_small_ele=smallindexele(arr,i);
            if(i!=index_small_ele)
            {
                int temp=arr[i];
                arr[i]=arr[index_small_ele];
                arr[index_small_ele]=temp;
            }
            
        }
    }
    public static int smallindexele(int arr[],int i)
    {
        int small_index=i;
        for(int j=i+1;j<arr.length;j++)
        {
            if(arr[small_index]>arr[j])
            {
                small_index=j;
            }
        }
        return small_index;
    }
}

// Time complexity of selection sort is depends on how many comparison durin each iteration
// so for each iterantion comparison of n-1, n-2,n-3........2
// so total time to comparison = n(n+1)/2-1
//                             =n(n+1)/2
//                             =n^2+n
//                             =n^2
//  so worst case time complexity = o(n^2)
//  Space complexity 
// total number of space complexity=o(1)
// total number of swapping approximately=n-1 
// total number of comparison =n^2;

File: Sorting_Algorithm/Selection_Sort/selectionsort.js
function selectionSort(arr)
{
    let n=arr.length;
    for(let i=0;i<n-1;i++)
    {
        let min_ele_index=minEleIndex(arr,i);
        if(min_ele_index!=i)
        {
            let temp=arr[min_ele_index];
            arr[min_ele_index]=arr[i];
            arr[i]=temp;
        }
    }
}

function minEleIndex(arr,i)
{
    let index_min_ele=i;
    for(let j=i+1;j<arr.length;j++)
    {
        if(arr[index_min_ele]>arr[j]){
            index_min_ele=j;
        }
    }
    return index_min_ele;
}

 const arr=[9,6,2,1,4] //Test Case 1
// const arr=[3,-2,1,0,5,12] //Test Case 2
// const arr=[45,12] //Test Case 3
// const arr=[-3,-1,-5] //Test Case 4
// const arr=["Nilesh","Ridant","Jyoti"] //Test Case 5
// const arr=[12,"NSS",45,-12,45.7] //Test Case 6
//const arr=[34,12,5,"NSS",23.2] //Test Case 7

console.log("Input array=",arr)
selectionSort(arr)
console.log("sorted array using selection sort algorithm=","\n",arr)
